File: _prologue/server/entity.go
/* */

package main

import (
	"roguelike-online/_prologue/server/components"
	"strconv"
)

type entity struct {
	Id                    int
	Name                  string `json:"name"`
	Kind                  string `json:"kind"`
	components.Position   `json:"pos"`
	components.Properties `json:"props"`
}

func (e *entity) isBlocking() bool {
	return e.Blocks
}

func (e *entity) isMobile() bool {
	return e.Mobile
}

func (e *entity) move(action string, m zoneMap, es entities) bool {
	switch action {
	case "UP":
		e.Target.Y--
	case "UPRIGHT":
		e.Target.X++
		e.Target.Y--
	case "RIGHT":
		e.Target.X++
	case "DOWNRIGHT":
		e.Target.X++
		e.Target.Y++
	case "DOWN":
		e.Target.Y++
	case "DOWNLEFT":
		e.Target.X--
		e.Target.Y++
	case "LEFT":
		e.Target.X--
	case "UPLEFT":
		e.Target.X--
		e.Target.Y--
	}
	if !es.isPointEmpty(e.Target) {
		e.Target = e.Current
		return false
	}
	isDiagonal := sliceContainsString(action, diagonalMovements)
	if m.isWalkableP(e.Target) && !isDiagonal {
		e.Current = e.Target
		return true
	}
	if isDiagonal && e.canMoveDiagonal(action, m) && m.isWalkableP(e.Target) {
		e.Current = e.Target
		return true
	}
	e.Target = e.Current
	return false
}

func (e *entity) canMoveDiagonal(action string, m zoneMap) bool {
	switch action {
	case "UPRIGHT":
		if !m.isWalkable(e.Current.X, e.Current.Y-1) {
			return false
		}
		if !m.isWalkable(e.Current.X+1, e.Current.Y) {
			return false
		}
	case "DOWNRIGHT":
		if !m.isWalkable(e.Current.X, e.Current.Y+1) {
			return false
		}
		if !m.isWalkable(e.Current.X+1, e.Current.Y) {
			return false
		}
	case "DOWNLEFT":
		if !m.isWalkable(e.Current.X, e.Current.Y+1) {
			return false
		}
		if !m.isWalkable(e.Current.X-1, e.Current.Y) {
			return false
		}
	case "UPLEFT":
		if !m.isWalkable(e.Current.X, e.Current.Y-1) {
			return false
		}
		if !m.isWalkable(e.Current.X-1, e.Current.Y) {
			return false
		}
	}
	return true
}

func newEntity(p components.Point, counter int) *entity {
	x := p.X
	y := p.Y
	name := "name" + "_" + strconv.Itoa(counter)
	kind := "rat"
	e := &entity{
		Id:   counter,
		Name: name,
		Kind: kind,
		Position: components.Position{
			Current: components.Point{
				X: x,
				Y: y,
			},
			Target: components.Point{
				X: x,
				Y: y,
			},
			View: components.Point{
				X: -1,
				Y: -1,
			},
		},
		Properties: components.Properties{
			Blocks: true,
			Mobile: true,
		},
	}
	return e
}

File: _prologue/core/entity.js
/* */

console.log('Loading...../core/entity.js');

import { K } from "./_konfig.js";
import { utils as u } from "./utils.js";
import { Point } from "./utils.js";
import { r } from "./run.js";
import { entities } from "./entities.js";

function createEntity(id, type, pos) {
  return new Entity(id, type, pos);
}

class Entity {
  constructor(id, type, pos) {
    this.id = id;
    this.name = type + "_" + id;
    this.type = type;
    this.start(pos);
  }
  start(pos) {
    this.pos = pos;// new Point(pos.x, pos.y);
    this.view = pos; //new Point(x, y);
  }
  move(action) {
    const target = new Point(this.pos.x, this.pos.y);
    switch (action) {
      case "SKIP":
        return true;
      case "UPLEFT":
        target.x -= 1;
        target.y -= 1;
        break;
      case "UP":
        target.y -= 1;
        break;
      case "UPRIGHT":
        target.x += 1;
        target.y -= 1;
        break;
      case "RIGHT":
        target.x += 1;
        break;
      case "DOWNRIGHT":
        target.x += 1;
        target.y += 1;
        break;
      case "DOWN":
        target.y += 1;
        break;
      case "DOWNLEFT":
        target.x -= 1;
        target.y += 1;
        break;
      case "LEFT":
        target.x -= 1;
        break;
      default:
        break;
    }

    if (this.canMove(target)) {
      this.pos = target;
      return true;
    }
    return false;
  }
  canMove(p) {
    const tile = r.map[p.x][p.y];
    if (!tile.walkable) {
      return false;
    }
    if (!entities.isEmptyPoint(p, r.entities)) {
      return false;
    }
    return true;
  }
}

export {
  createEntity,
};
